talk/expert/cpp20spaceship.tex

```diff
@@ -16,216 +24,141 @@
-  \frametitlecpp[20]{The three-way comparison operator}
-  \scriptsize
-  \begin{exampleblock}{Example}
-    \begin{cppcode*}{}
-    template <typename T>
-    void compare( T lhs, T rhs )
-    {
-      std::cout<<lhs<<" <=> "<<rhs<<" (< <= > >=) : "<<(lhs<=>rhs)
-        <<" ("<<(lhs<rhs)<<" "<<(lhs<=rhs)
-        <<" "<<(lhs>rhs)<<" "<<(lhs>=rhs)<<")"
-        <<std::endl ;
+      // Ratio uses weak_ordering
+      three_way_compare(Ratio{1, 2}, Ratio{2, 3}); // (1/2<=>2/3) : 100
+      three_way_compare(Ratio{2, 3}, Ratio{4, 6}); // (2/3<=>4/6) : 010
+      three_way_compare(Ratio{2, 3}, Ratio{1, 2}); // (2/3<=>1/2) : 001
     }
-    int main()
-    {
-      compare(Ratio{ 3, 4 }, Ratio{ 2, 3 }) ;
-      compare(Ratio{ 3, 6 }, Ratio{ 2, 3 }) ;
-      compare(Ratio{ 1, 2 }, Ratio{ 2, 4 }) ;
-    }
-     \end{cppcode*}
+    \end{cppcode*}
   \end{exampleblock}
-  \begin{Verbatim}
-3/4 <=> 2/3 (< <= > >=) : 001 (0 0 1 1)
-3/6 <=> 2/3 (< <= > >=) : 100 (1 1 0 0)
-1/2 <=> 2/4 (< <= > >=) : 010 (0 1 0 1)
-  \end{Verbatim}
 \end{frame}
 
 \begin{frame}[fragile]
-  \frametitlecpp[20]{The three-way comparison operator}
-  \begin{block}{Home-made class}
-    \begin{itemize}
-      \item As one can see from previous code, we have defined only \mintinline{cpp}{<=>}, but \mintinline{cpp}{<}, \mintinline{cpp}{>}, \mintinline{cpp}{<=} and \mintinline{cpp}{=>} work as well. The compiler can express the latter operators in terms of \mintinline{cpp}{<=>}. Of course, one can also provide own implementations.
-      \item It will \emph{NOT} add a default implementation for \mintinline{cpp}{==} and \mintinline{cpp}{!=}. Those operators are generally expected to mean \textbf{equal}, rather than \textbf{equivalent}. If \mintinline{cpp}{<=>} does not provide a strong order, it is generally advised not to define \mintinline{cpp}{==}.
-    \end{itemize}
-  \end{block}
-\end{frame}
-
-\begin{frame}[fragile]
-  \frametitlecpp[20]{The three-way comparison operator}
-  \begin{block}{Default \texttt{<=>} implementation}
+  \frametitlecpp[20]{Compiler generated comparison operators}
+  \begin{block}{For a given home-made class}
     \begin{itemize}
-      \item One can ask the compiler to provide a default implementation for \mintinline{cpp}{<=>} and/or \mintinline{cpp}{==}. Logically enough, it will compare the first member variable of the two objects, and goes on to the next member variable as long as the current ones are equivalent.
-      \item In the previous example, that would be wrong, because it will compare the numerators first, and conclude that \mintinline{cpp}{3/6} is greater than \mintinline{cpp}{2/3}:
-      \item On the contrary, for some tuple-like class like the next example, it makes sense.
+    \item defining \mintinline{cpp}{operator<=>} allows the compiler to generate \mintinline{cpp}{operator<}, \mintinline{cpp}{operator>}, \mintinline{cpp}{operator<=} and \mintinline{cpp}{operator=>}
+    \item of course, one can overwrite the default implementations
```

Comment:
Not sure if this word is better. Feel free to disregard this.
```suggestion
    \item of course, one can replace the default implementations
```

Reply:
In a previous debate, we emphasized  that the compiler does not really generate < (and > and...) in the class. Rather, each time it encounters a use of < in the code, it will rephrase it in terms of <=> (TO BE CONFIRMED). This can be seen as a detail, but perhaps it may help to known this when trying to decipher a compilation error involving those operators.

Reply:
@chavid, I've replaced the generate line with "allows the compiler to use it when encountering other comparison operators", so that we stay vague enough